### PR TITLE
core: Move custom Operation verification 'verify_()' to end of verify function

### PR DESCRIPTION
**Pull request by ed741**
At the moment the verification process for an Operation looks like this:

    def verify(self, verify_nested_ops: bool = True) -> None:
        for operand in self.operands:
            if isinstance(operand, ErasedSSAValue):
                raise Exception("Erased SSA value is used by the operation")

        # Custom verifier
        self.verify_()

        # Verifier generated by irdl_op_def
        op_def = type(self).irdl_definition
        if op_def is not None:
            op_def.verify(self)

        if verify_nested_ops:
            for region in self.regions:
                region.verify()

    def verify_(self) -> None:
        pass
By doing the Custom Verifer before the irdl generated varifer when an attribute is missed in the parsed input text (for example), it becomes an ugly error when it is assumed to exist in the  ``` verify_()->None ``` function:
```
... xdsl/xdsl/irdl.py", line 1179, in field_getter
    return self.attributes[attribute_name]
KeyError: 'opAttr_var_name'
```
whereas if the irdl generated verifier went first we get the slightly nicer error:
```
xdsl.utils.exceptions.VerifyException: attribute opAttr_var_name expected
```
And there would be room, I assume, to make this error prettier in the future, without users needing to worry about performing the same verification checks in their custom  ```verify_()```  functions.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]
Good idea, I don't see any issues with it, and actually already assumed that the custom verifier was run after the generated ones. Let's see if any tests fail, and if @math-fehr and @webmiche agree

File: xdsl/ir.py [hunk 1]
```diff
@@ -689,6 +686,9 @@ def verify(self, verify_nested_ops: bool = True) -> None:
         if verify_nested_ops:
             for region in self.regions:
                 region.verify()
+
```
Comment (superlopuh) [inline xdsl/ir.py:689]:
```suggestion

```

**Review by math-fehr — APPROVED** [review]
Yes, this was a bug in our code, and IRDL should always be run before the custom verifier!
Note that we should probably add a test for that, so it won't happen again!

File: tests/test_ir.py [hunk 2]
```diff
@@ -286,3 +286,24 @@ def test_descriptions():
 \t  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
 \t}
 )'''
+
+
+def test_op_custom_verify_is_done_last():
+    a = Constant.from_int_and_width(1, i32)
```
Comment (superlopuh) [inline tests/test_ir.py:292]: I'd recommend a different test, maybe something like this:
``` python
@irdl_op_def
class CustomVerifyLastOp(Operation):
    name = 'test.custom_verify_last_op'
    val: Annotated[Operand, i64]

    @staticmethod
    def get(val: SSAValue):
        return CustomVerifyLastOp.build(operands: [val])

    def verify_(self):
        raise VerificationError('should not be triggered')

def test_op_custom_verify_is_done_last():
    a = Constant.from_int_and_width(1, i32)
    b = CustomVerifyLastOp.get(a.res)
    with pytest.raises(VerificationError) as e:
        b.verify
    assert e.value.args[0] == 'I forget what the message is supposed to be here'
```

**Review by superlopuh — CHANGES_REQUESTED** [review]
would rework the test a bit, it feels like it's not testing the built-in functions themselves

File: tests/test_ir.py (hunk 2 above)
Comment (math-fehr) [reply tests/test_ir.py:292]: Agreed!

Comment (superlopuh) [reply tests/test_ir.py:292]: The janky system is also coupled to the implementation. At the end of the day, we're testing irdl, so it feels like representative code that a user might write is best.

**Review by superlopuh — APPROVED** [review]
Perfect, thank you

**Review by webmiche — APPROVED** [review]
LGTM

**Pull request merged**